### PR TITLE
Fix Bug #33478: Add FP16/BF16 precision support to Math node

### LINKED ISSUE #33478: [Bug]: Significant Numerical Inconsistency with Softmax -> Cos -> ArgMax sequence on CPU (FP16)
### OpenVINO Version

openvino  2025.4.0

### Operating System

Ubuntu 20.04 (LTS)

### Device used for inference

CPU

### Framework

None

### Model used

_No response_

### Issue description

I found a significant numerical inconsistency between PyTorch and OpenVINO on CPU when running a specific sequence of operators in FP16 precision.

The graph logic computes ArgMax(Cos(Softmax(x))). Since Cos is monotonic in the range [0, 1] (output of Softmax), this logic is effectively finding extrema indices. However, OpenVINO produces a completely different integer sum compared to PyTorch (Reference).

- PyTorch Result: 19 (Expected)
- OpenVINO Result: 727 (Incorrect)
- Difference: 708

This suggests a logical error or precision collapse in the CPU plugin's optimization pass when fusing or executing these operators in FP16.

### Run the following reproduction script.

import torch
import numpy as np
import openvino.runtime as ov
import onnx

#1. Define the problematic sub-graph
class BugModel(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.softmax = torch.nn.Softmax(dim=-1) 

    def forward(self, x):
        # Input -> Softmax -> Cos -> ArgMax(dim=0) -> ReduceSum(dim=0)
        v4_0 = self.softmax(x)
        v3_0 = torch.cos(v4_0)
        v1_0 = torch.argmax(v3_0, dim=0)
        v2_0 = v1_0.sum(dim=0)
        return v2_0

#2. Prepare Data (FP16 is key to reproduction)
#Shape derived from fuzzing report
input_shape = (27, 59)
#Fixed seed for reproducibility
np.random.seed(42)
x_numpy = np.random.randn(*input_shape).astype(np.float16)
x_torch = torch.tensor(x_numpy)

#3. Run PyTorch (Oracle)
model = BugModel()
model.eval()
with torch.no_grad():
    y_torch = model(x_torch)
    print(f"PyTorch Result: {y_torch.item()}")

#4. Export to ONNX
onnx_path = "bug_repro.onnx"
torch.onnx.export(
    model, 
    x_torch, 
    onnx_path, 
    input_names=['v0_0'], 
    output_names=['v2_0'],
    opset_version=14
)
print(f"Model exported to {onnx_path}")

#5. Run OpenVINO
core = ov.Core()
ov_model = core.read_model(onnx_path)
compiled_model = core.compile_model(ov_model, "CPU")
infer_request = compiled_model.create_infer_request()

#Inference
res = infer_request.infer([x_numpy])[compiled_model.output(0)]
y_ov = res

print(f"OpenVINO Result: {y_ov}")

#6. Compare
diff = np.abs(y_torch.numpy() - y_ov)
print(f"Difference: {diff}")

### Relevant log output

```shell
PyTorch Result: 19
Model exported to bug_repro.onnx
OpenVINO Result: 727
Difference: 708
❌ Bug Reproduced! Significant difference detected.
```

### Issue submission checklist

- [x] I'm reporting an issue. It's not a question.
- [x] I checked the problem with the documentation, FAQ, open issues, Stack Overflow, etc., and have not found a solution.
- [x] There is reproducer code and related data files such as images, videos, models, etc.

### PR DESCRIPTION
**Pull request by vishalharkal15**
The Math node was forcing F32 precision for all operations, causing numerical inconsistencies when FP16 inference was enabled. This resulted in significant errors (e.g., 727 instead of 19) in the Softmax->Cos->ArgMax sequence.

Changes:
- Updated initSupportedPrimitiveDescriptors() to respect input precision (F32/BF16/F16) instead of hardcoding F32
- Added templated executeImpl<T>() to support multiple precisions
- Updated execute() to dispatch based on actual data precision
- Added required headers for float16 and bfloat16 types
- Migrated from deprecated cpu_parallel API to parallel_nt()

Fixes #33478

### Details:
 - *item1*
 - *...*

### Tickets:
 - *ticket-id*

### CODE REVIEW

**Review by vishalharkal15 — COMMENTED** [review]
Fix sanitizer issues: Use cpu_parallel API instead of parallel_nt
The previous implementation used parallel_nt() which caused issues with
Linux sanitizers. This commit:
- Reverts to using cpu_parallel->parallel_for() from GraphContext
- Keeps FP16/BF16 precision support via template executeImpl<T>()
- Removes unnecessary openvino/core/parallel.hpp include
- Adds executeImpl<T>() method declaration to header file

This maintains the fix for Bug https://github.com/openvinotoolkit/openvino/issues/33478 (FP16 precision support) while
using the proper thread-safe parallel execution API.
